In [ ]:
import pandas as pd

In [27]:
data = pd.read_csv("clean_data.csv")
data.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


In [28]:
# count the number of unique customers in dataset

data["CustomerID"].nunique()

4338

In [29]:
data["Revenue"] = data["Quantity"] * data["UnitPrice"]

data.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,Revenue
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom,15.30
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom,22.00
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34


In [30]:
# What is the total revenue?

data["Revenue"].sum()

np.float64(8887208.894000001)

In [31]:
total_revenue = data["Revenue"].sum()

print(round(total_revenue, 2))

8887208.89


The cleaned dataset contains 4338 unique customers and generated approximately 8.89M pounds during the period analysed.

In [32]:
# group the data by country and calc total revenue for each country

revenue_by_country = data.groupby("Country")["Revenue"].sum()

# sort
revenue_by_country = revenue_by_country.sort_values(ascending=False)

revenue_by_country

Country
United Kingdom          7285024.644
Netherlands              285446.340
EIRE                     265262.460
Germany                  228678.400
France                   208934.310
Australia                138453.810
Spain                     61558.560
Switzerland               56443.950
Belgium                   41196.340
Sweden                    38367.830
Japan                     37416.370
Norway                    36165.440
Portugal                  33375.840
Finland                   22546.080
Singapore                 21279.290
Channel Islands           20440.540
Denmark                   18955.340
Italy                     17483.240
Cyprus                    13502.850
Austria                   10198.680
Poland                     7334.650
Israel                     7215.840
Greece                     4760.520
Iceland                    4310.000
Canada                     3666.380
USA                        3580.390
Malta                      2725.590
Unspecified         

In [33]:
uk_revenue = revenue_by_country["United Kingdom"]

uk_percentage = uk_revenue / total_revenue * 100

print(round(uk_percentage, 2))

81.97


Analysis: Revenue is heavily concentrated in the UK. International markets collectively contribute a much smaller share of revenue.

In [34]:
international_revenue = revenue_by_country[revenue_by_country.index != "United Kingdom"]

international_revenue.head(5)

Country
Netherlands    285446.34
EIRE           265262.46
Germany        228678.40
France         208934.31
Australia      138453.81
Name: Revenue, dtype: float64

In [35]:
# convert data type of InvoiceDate from string to datetime

data["InvoiceDate"] = pd.to_datetime(data["InvoiceDate"])
data["InvoiceDate"].dtype
data["InvoiceDate"].head()

0   2010-12-01 08:26:00
1   2010-12-01 08:26:00
2   2010-12-01 08:26:00
3   2010-12-01 08:26:00
4   2010-12-01 08:26:00
Name: InvoiceDate, dtype: datetime64[us]

In [36]:
# create new column for Month

data["Month"] = data["InvoiceDate"].dt.to_period("M")

data.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,Revenue,Month
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom,15.30,2010-12
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34,2010-12
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom,22.00,2010-12
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34,2010-12
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34,2010-12


In [37]:
# group data by month and calc total revenue for each month

revenue_by_month = data.groupby("Month")["Revenue"].sum()

revenue_by_month

Month
2010-12     570422.730
2011-01     568101.310
2011-02     446084.920
2011-03     594081.760
2011-04     468374.331
2011-05     677355.150
2011-06     660046.050
2011-07     598962.901
2011-08     644051.040
2011-09     950690.202
2011-10    1035642.450
2011-11    1156205.610
2011-12     517190.440
Freq: M, Name: Revenue, dtype: float64

Note: Earlier in the other notebook -> Earliest dated txn: Dec 1 2010, Latest dated txn: Dec 9 2021
So 517190.440 represents the revenue for Dec 2021 as of 9 Dec (9 day revenue).

Analysis: Monthly revenue remained relatively stable through early and mid-2011 before increasing substantially from September onwards. Revenue reached approx. £1.16M in November 2011, the highest full month in the dataset and roughly 80% higher than August 2011. 

This suggests higher purchasing activity toward the year-end period in Q4 potentially due to holiday/gifting season. December 2011 was excluded from month-to-month interpretation because the dataset only contains transactions through 9 December.

In [38]:
# calculate number of unique orders for each month

orders_by_month = data.groupby("Month")["InvoiceNo"].nunique()

orders_by_month

Month
2010-12    1400
2011-01     987
2011-02     997
2011-03    1321
2011-04    1149
2011-05    1555
2011-06    1393
2011-07    1331
2011-08    1280
2011-09    1755
2011-10    1929
2011-11    2657
2011-12     778
Freq: M, Name: InvoiceNo, dtype: int64

It seems the year-end surge in revenue was substantially driven by an incease in number of orders.

In [40]:
aug_revenue = revenue_by_month["2011-08"]
nov_revenue = revenue_by_month["2011-11"]

revenue_uplift = (nov_revenue - aug_revenue) / aug_revenue * 100

print("Revenue uplift:", round(revenue_uplift, 2), "%")

aug_orders = orders_by_month["2011-08"]
nov_orders = orders_by_month["2011-11"]

order_uplift = (nov_orders - aug_orders) / aug_orders * 100

print("Order uplift:", round(order_uplift, 2), "%")

Revenue uplift: 79.52 %
Order uplift: 107.58 %


Orders grew faster than revenue. So the number of orders increased by the ABS fell during this period (Aug - Nov).

In [41]:
# calculate average order value for each month

average_order_value = revenue_by_month / orders_by_month

average_order_value

Month
2010-12    407.444807
2011-01    575.583901
2011-02    447.427202
2011-03    449.721241
2011-04    407.636493
2011-05    435.598167
2011-06    473.830617
2011-07    450.009693
2011-08    503.164875
2011-09    541.703819
2011-10    536.880482
2011-11    435.154539
2011-12    664.769203
Freq: M, dtype: float64

In [43]:
aug_abs = average_order_value["2011-08"]
nov_abs = average_order_value["2011-11"]

abs_change = (nov_abs - aug_abs) / aug_abs * 100

print("August ABS: £", round(aug_abs, 2))
print("November ABS: £", round(nov_abs, 2))
print("ABS change:", round(abs_change, 2), "%")

August ABS: £ 503.16
November ABS: £ 435.15
ABS change: -13.52 %


Revenue growth from August to November was primarily driven by increased order volume rather than higher spending per order. While we observe an order uplift of 108% (no. of orders more than doubled), ABS declined by about 14% over the same period.

Conclusions:

Customer base: 4338

Total revenue: £8.89M

UK share of revenue: 82%

Trends: Strong revenue observed in Q4 months, driven by higher order volume

In [44]:
# calculate number of orders made by each customer

frequency = data.groupby("CustomerID")["InvoiceNo"].nunique()

frequency

CustomerID
12346.0     1
12347.0     7
12348.0     4
12349.0     1
12350.0     1
           ..
18280.0     1
18281.0     1
18282.0     2
18283.0    16
18287.0     3
Name: InvoiceNo, Length: 4338, dtype: int64

In [ ]:
# sort customers based on frequency of purchases

frequency = frequency.sort_values(ascending=False)

frequency.head(10)

CustomerID
12748.0    209
14911.0    201
17841.0    124
13089.0     97
14606.0     93
15311.0     91
12971.0     86
14646.0     73
16029.0     63
13408.0     62
Name: InvoiceNo, dtype: int64

In [46]:
# calculate total revenue generated by each customer

monetary = data.groupby("CustomerID")["Revenue"].sum()

monetary

CustomerID
12346.0    77183.60
12347.0     4310.00
12348.0     1797.24
12349.0     1757.55
12350.0      334.40
             ...   
18280.0      180.60
18281.0       80.82
18282.0      178.05
18283.0     2045.53
18287.0     1837.28
Name: Revenue, Length: 4338, dtype: float64

In [ ]:
monetary = monetary.sort_values(ascending=False)

# top 10 customers in spending
monetary.head(10)

CustomerID
14646.0    280206.02
18102.0    259657.30
17450.0    194390.79
16446.0    168472.50
14911.0    143711.17
12415.0    124914.53
14156.0    117210.08
17511.0     91062.38
16029.0     80850.84
12346.0     77183.60
Name: Revenue, dtype: float64

Observations: Customer 14646 ranks 8th in terms of purchase frequency (73 orders), however they are the top user in terms of monetary spending. 

Insight: Most frequent customer != most valuable customer

In [48]:
# Recency

# find the latest purchase date for each customer

last_purchase = data.groupby("CustomerID")["InvoiceDate"].max()

last_purchase

CustomerID
12346.0   2011-01-18 10:01:00
12347.0   2011-12-07 15:52:00
12348.0   2011-09-25 13:13:00
12349.0   2011-11-21 09:51:00
12350.0   2011-02-02 16:01:00
                  ...        
18280.0   2011-03-07 09:52:00
18281.0   2011-06-12 10:53:00
18282.0   2011-12-02 11:43:00
18283.0   2011-12-06 12:02:00
18287.0   2011-10-28 09:29:00
Name: InvoiceDate, Length: 4338, dtype: datetime64[us]

In [ ]:
# note the smaller no. of days, the more recent the purchase
# the more likely the customer is to return to make another purchase

# assume ana;ysis is performed on 10 dec, 1 day after the last purchase date in the dataset
reference_date = pd.to_datetime("2011-12-10")

print(reference_date)

2011-12-10 00:00:00


In [52]:
recency = (reference_date - last_purchase).dt.days

recency

CustomerID
12346.0    325
12347.0      2
12348.0     75
12349.0     18
12350.0    310
          ... 
18280.0    277
18281.0    180
18282.0      7
18283.0      3
18287.0     42
Name: InvoiceDate, Length: 4338, dtype: int64

In [ ]:
# I'm curious about the customer with ID 14646 (from earlier), who is the top spender in the dataset
# want to see if they also purchased recently

print("Recency:", recency[14646])
print("Frequency:", frequency[14646])
print("Monetary: £", round(monetary[14646], 2))

Recency: 1
Frequency: 73
Monetary: £ 280206.02


Analysis: Customer 14646 is the highest-spending customer in the dataset, generating approx. £280K in revenue. The customer also made 73 unique txns and purchased within 1 day of the reference date, suggesting that they are not only high-value but also highly frequent and recently active.

In [ ]:
# combine recency, frequency and monetary into one table
rfm = pd.concat([recency, frequency, monetary], axis=1)

# rename the columns because we prev grouped them by 
# the invoice date (R), number of unique orders (F) and revenues (M)
rfm.columns = ["Recency", "Frequency", "Monetary"]

rfm.head(10)

,Recency,Frequency,Monetary
CustomerID,,,
14441.0,0,4,1712.89
14446.0,0,4,1010.35
16446.0,0,2,168472.50
15804.0,0,13,4206.39
17364.0,0,11,4462.68
16705.0,0,20,14034.99
12662.0,0,11,3849.78
16626.0,0,17,4413.10
18102.0,0,60,259657.30


In [60]:
rfm.tail(10)

,Recency,Frequency,Monetary
CustomerID,,,
12791.0,373,1,192.60
14729.0,373,1,313.49
18074.0,373,1,489.60
16048.0,373,2,256.44
17643.0,373,1,101.55
16583.0,373,1,233.45
17968.0,373,1,265.10
17908.0,373,1,232.03
14142.0,373,1,311.81


### RFM SCORING

In [ ]:
# R scoring: divide customers into 5 groups based on recency
# 5 = most recent customers, 1 = least recent customers

rfm["R_Score"] = pd.qcut(
    rfm["Recency"],
    5,
    labels=[5, 4, 3, 2, 1]
)

rfm.head(10)

,Recency,Frequency,Monetary,R_Score
CustomerID,,,,
14441.0,0,4,1712.89,5
14446.0,0,4,1010.35,5
16446.0,0,2,168472.50,5
15804.0,0,13,4206.39,5
17364.0,0,11,4462.68,5
16705.0,0,20,14034.99,5
12662.0,0,11,3849.78,5
16626.0,0,17,4413.10,5
18102.0,0,60,259657.30,5


In [ ]:
# rfm["F_Score"] = pd.qcut(
    # rfm["Frequency"],
    # 5,
    # labels=[1, 2, 3, 4, 5]
# )

# code above failed because there are many customers with the same frequency value, so the qtys cannot be calculated properly. 
# Instead, we can use value_counts to see the distribution of frequency values.

rfm["Frequency"].value_counts().sort_index()

Frequency
1      1493
2       835
3       508
4       388
5       242
6       172
7       143
8        98
9        68
10       54
11       52
12       45
13       30
14       20
15       28
16       11
17       18
18       14
19       12
20       12
21       11
22        5
23        5
24        3
25        8
26        7
27        3
28        6
29        1
30        4
31        3
32        3
33        2
34        3
35        1
37        3
38        2
39        2
41        1
44        1
45        1
46        1
47        2
48        1
50        1
51        1
55        2
57        1
60        1
62        1
63        1
73        1
86        1
91        1
93        1
97        1
124       1
201       1
209       1
Name: count, dtype: int64

Analysis: Frequency data is very uneven. Out of 4338 customers, 1493 only made one unique order (34%). This makes it difficult to divide customers into equal-sized groups, so I will use custom frequency bins instead based on purchase behaviour.

In [67]:
def frequency_score(frequency):

    if frequency == 1:
        return 1

    elif frequency == 2:
        return 2

    elif frequency <= 4:
        return 3

    elif frequency <= 9:
        return 4

    else:
        return 5

rfm["F_Score"] = rfm["Frequency"].apply(frequency_score)

rfm.head(10)

,Recency,Frequency,Monetary,R_Score,F_Score
CustomerID,,,,,
14441.0,0,4,1712.89,5,3
14446.0,0,4,1010.35,5,3
16446.0,0,2,168472.50,5,2
15804.0,0,13,4206.39,5,5
17364.0,0,11,4462.68,5,5
16705.0,0,20,14034.99,5,5
12662.0,0,11,3849.78,5,5
16626.0,0,17,4413.10,5,5
18102.0,0,60,259657.30,5,5


In [68]:
# M scoring
# 5 = highest spending customers, 1 = lowest spending customers

rfm["M_Score"] = pd.qcut(
    rfm["Monetary"],
    5,
    labels=[1, 2, 3, 4, 5]
)

rfm.head(10)

,Recency,Frequency,Monetary,R_Score,F_Score,M_Score
CustomerID,,,,,,
14441.0,0,4,1712.89,5,3,4
14446.0,0,4,1010.35,5,3,4
16446.0,0,2,168472.50,5,2,5
15804.0,0,13,4206.39,5,5,5
17364.0,0,11,4462.68,5,5,5
16705.0,0,20,14034.99,5,5,5
12662.0,0,11,3849.78,5,5,5
16626.0,0,17,4413.10,5,5,5
18102.0,0,60,259657.30,5,5,5


In [69]:
# check how many customers received each score

print("R Scores:")
print(rfm["R_Score"].value_counts().sort_index())

print("\nF Scores:")
print(rfm["F_Score"].value_counts().sort_index())

print("\nM Scores:")
print(rfm["M_Score"].value_counts().sort_index())

R Scores:
R_Score
5    868
4    880
3    863
2    866
1    861
Name: count, dtype: int64

F Scores:
F_Score
1    1493
2     835
3     896
4     723
5     391
Name: count, dtype: int64

M Scores:
M_Score
1    868
2    867
3    868
4    867
5    868
Name: count, dtype: int64


Analysis: Customers are relatively evenly distributed across the Recency and Monetary scores. Frequency is more uneven because a large proportion of customers made only one or a few purchases, so custom bins were used to better reflect actual purchasing behaviour.

## CUSTOMER SEGMENTATION

In [ ]:
# calculate total RFM score for each customer
# convert to type int

rfm["RFM_Score"] = (
    rfm["R_Score"].astype(int)
    + rfm["F_Score"].astype(int)
    + rfm["M_Score"].astype(int)
)

rfm.head(10)

,Recency,Frequency,Monetary,R_Score,F_Score,M_Score,RFM_Score
CustomerID,,,,,,,
14441.0,0,4,1712.89,5,3,4,12
14446.0,0,4,1010.35,5,3,4,12
16446.0,0,2,168472.50,5,2,5,12
15804.0,0,13,4206.39,5,5,5,15
17364.0,0,11,4462.68,5,5,5,15
16705.0,0,20,14034.99,5,5,5,15
12662.0,0,11,3849.78,5,5,5,15
16626.0,0,17,4413.10,5,5,5,15
18102.0,0,60,259657.30,5,5,5,15


In [73]:
# create customer segments based on RFM scores

def customer_segment(row):

    # if customer purchsed recently, purchases frequently and spends a lot
    if row["R_Score"] >= 4 and row["F_Score"] >= 4 and row["M_Score"] >= 4:
        return "High-Value Loyal"

    # if customer purchased recently, but hasn't purchased frequently
    elif row["R_Score"] >= 4 and row["F_Score"] <= 2:
        return "Recent / Low Frequency"

    # if customer hasn't purchased recently, but historically purchased frequently
    elif row["R_Score"] <= 2 and row["F_Score"] >= 4:
        return "Risk of Churn"

    # if customer hasn't purchased recently and hasn't purchased frequently
    elif row["R_Score"] <= 2 and row["F_Score"] <= 2:
        return "Inactive / Low Frequency"
    
    else:
        return "Regular"

In [74]:
rfm["Segment"] = rfm.apply(customer_segment, axis=1)

rfm.head(10)

,Recency,Frequency,Monetary,R_Score,F_Score,M_Score,RFM_Score,Segment
CustomerID,,,,,,,,
14441.0,0,4,1712.89,5,3,4,12,Regular
14446.0,0,4,1010.35,5,3,4,12,Regular
16446.0,0,2,168472.50,5,2,5,12,Recent / Low Frequency
15804.0,0,13,4206.39,5,5,5,15,High-Value Loyal
17364.0,0,11,4462.68,5,5,5,15,High-Value Loyal
16705.0,0,20,14034.99,5,5,5,15,High-Value Loyal
12662.0,0,11,3849.78,5,5,5,15,High-Value Loyal
16626.0,0,17,4413.10,5,5,5,15,High-Value Loyal
18102.0,0,60,259657.30,5,5,5,15,High-Value Loyal


In [75]:
# count number of customers in each segment

segment_counts = rfm["Segment"].value_counts()

segment_counts

Segment
Regular                     1593
Inactive / Low Frequency    1335
High-Value Loyal             789
Recent / Low Frequency       515
Risk of Churn                106
Name: count, dtype: int64

In [77]:
# calculate percentage of customers in each segment

segment_percentage = segment_counts / len(rfm) * 100

segment_percentage.round(2)

Segment
Regular                     36.72
Inactive / Low Frequency    30.77
High-Value Loyal            18.19
Recent / Low Frequency      11.87
Risk of Churn                2.44
Name: count, dtype: float64

Analysis: Regular customers form the largest segment at 36.72% of the customer base. However, 30.77% of customers are inactive and have historically purchased infrequently, suggesting a sizeable portion of the customer base has relatively low engagement. 

Meanwhile, 11.87% are recent but low-frequency customers, presenting an opportunity to encourage repeat purchases. High-Value Loyal customers account for 18.19% and represent an important segment to retain and reward. Only 2.44% are classified as Risk of Churn, representing previously frequent customers who have not purchased recently.

In [78]:
# calculate total revenue generated by each customer segment

revenue_by_segment = rfm.groupby("Segment")["Monetary"].sum()

# sort from highest to lowest revenue

revenue_by_segment = revenue_by_segment.sort_values(ascending=False)

revenue_by_segment

Segment
High-Value Loyal            5444730.170
Regular                     2093036.352
Inactive / Low Frequency     681943.022
Recent / Low Frequency       421024.830
Risk of Churn                246474.520
Name: Monetary, dtype: float64

High-Value Loyal customers generate about £5.44M, even though they are only 18.19% of customers. Meanwhile, Inactive / Low Frequency customers are 30.77% of customers, but generate only about £682K.

In [79]:
# calculate total revenue across all segments

total_segment_revenue = revenue_by_segment.sum()

# calculate percentage of revenue generated by each segment

revenue_percentage = revenue_by_segment / total_segment_revenue * 100

revenue_percentage.round(2)

Segment
High-Value Loyal            61.26
Regular                     23.55
Inactive / Low Frequency     7.67
Recent / Low Frequency       4.74
Risk of Churn                2.77
Name: Monetary, dtype: float64

Analysis: Revenue is highly concentrated among High-Value Loyal customers, who account for only 18.19% of the customer base but generate 61.26% of total revenue. 

In contrast, Inactive / Low Frequency customers represent 30.77% of customers but contribute only 7.67% of revenue. 

CRM strategies should prioritise retaining high-value loyal customers, while reactivation efforts for inactive customers may need to be more targeted. There also seems to be some opportunity to turn some regular customers into high-value, loyal customers. Based on prior findings, regular users make up 36.72% of total customers and generate 23.55% of revenue.

In [80]:
# turn all data into table for easier analysis

# create an empty table
segment_summary = pd.DataFrame()

# add number of customers
segment_summary["Customers"] = segment_counts

# add percentage of customers
segment_summary["% Customers"] = segment_percentage.round(2)

# add total revenue
segment_summary["Revenue"] = revenue_by_segment.round(2)

# add percentage of revenue
segment_summary["% Revenue"] = revenue_percentage.round(2)

segment_summary

,Customers,% Customers,Revenue,% Revenue
Segment,,,,
Regular,1593,36.72,2093036.35,23.55
Inactive / Low Frequency,1335,30.77,681943.02,7.67
High-Value Loyal,789,18.19,5444730.17,61.26
Recent / Low Frequency,515,11.87,421024.83,4.74
Risk of Churn,106,2.44,246474.52,2.77


It is important to concentrate strategies that target retaining high-value loyal customers (highly valuable customer segment). We should also identify high-value loyal customers at risk of churn.

In [81]:
# filter for customers who are at risk of churn

risk_customers = rfm[rfm["Segment"] == "Risk of Churn"]

risk_customers.head(10)

,Recency,Frequency,Monetary,R_Score,F_Score,M_Score,RFM_Score,Segment
CustomerID,,,,,,,,
16321.0,72,5,373.65,2,4,2,8,Risk of Churn
14978.0,72,5,1921.37,2,4,4,10,Risk of Churn
13901.0,72,5,2867.20,2,4,5,11,Risk of Churn
13397.0,72,5,1544.13,2,4,4,10,Risk of Churn
16681.0,72,6,1770.34,2,4,4,10,Risk of Churn
17656.0,73,5,1674.69,2,4,4,10,Risk of Churn
17061.0,73,11,5232.23,2,5,5,12,Risk of Churn
15674.0,73,5,2446.60,2,4,5,11,Risk of Churn
14101.0,73,6,5976.79,2,4,5,11,Risk of Churn


In [82]:
# sort at-risk customers from highest to lowest spending

risk_customers = risk_customers.sort_values(
    "Monetary",
    ascending=False
)

risk_customers.head(10)

,Recency,Frequency,Monetary,R_Score,F_Score,M_Score,RFM_Score,Segment
CustomerID,,,,,,,,
16180.0,100,8,10254.18,2,4,5,11,Risk of Churn
13093.0,275,8,7832.47,1,4,5,10,Risk of Churn
12980.0,157,9,7374.90,2,4,5,11,Risk of Churn
16745.0,86,17,7180.70,2,5,5,12,Risk of Churn
13027.0,113,6,6912.00,2,4,5,11,Risk of Churn
15939.0,89,15,6115.01,2,5,5,12,Risk of Churn
14101.0,73,6,5976.79,2,4,5,11,Risk of Churn
16553.0,163,12,5719.82,2,5,5,12,Risk of Churn
17850.0,372,34,5391.21,1,5,5,11,Risk of Churn


In [83]:
# calculate revenue from top 10 highest-spending at-risk customers

top_10_risk_revenue = risk_customers.head(10)["Monetary"].sum()

print("Top 10 at-risk customer revenue: £", round(top_10_risk_revenue, 2))

Top 10 at-risk customer revenue: £ 67989.31


In [84]:
# calculate total revenue from all at-risk customers

total_risk_revenue = risk_customers["Monetary"].sum()

# calculate percentage contributed by top 10

top_10_percentage = top_10_risk_revenue / total_risk_revenue * 100

print("Top 10 share of at-risk revenue:", round(top_10_percentage, 2), "%")

Top 10 share of at-risk revenue: 27.58 %


Analysis: The top 10 highest-spending at-risk customers account for 27.58% of total revenue generated by the Risk of Churn segment. This suggests that the value of at-risk customers is relatively concentrated among a small group. 

Reactivation efforts could and should prioritise these historically high-value customers.

## MARKETING RECOMMENDATIONS

Based on the RFM customer segmentation, different CRM strategies can be used for each customer segment to improve retention, reactivation and customer value.

### High-Value Loyal
#### Strategy: Retain & Reward

High-Value Loyal customers represent only 18.19% of customers but contribute 61.26% of total revenue, making them the most commercially valuable customer segment.

**Recommended actions:**
- Prioritise retention through loyalty rewards and exclusive benefits.
- Provide early access to new products or promotions.
- Personalise product recommendations based on previous purchases.
- Avoid relying excessively on large discounts, as these customers already demonstrate high purchase frequency and spending. There is no need to focus budget on discounts for this segment.

### Risk of Churn
#### Strategy: Winback

Risk of Churn customers:
106 customers | 2.44% of customers | 2.77% of revenue

Although RoC customers represent only 2.44% of the customer base, these customers have historically purchased frequently but have not purchased recently.

The top 10 highest-spending customers alone account for 27.58% of the segment's historical revenue, suggesting that reactivation efforts should be prioritised based on customer value.

**Recommended actions:**
- Target high-value at-risk customers with personalised win-back campaigns.
    - E.g. Segment the high-value at-risk customers into different categories based on purchase history (Fashion consumers, Lifestyle, Beauty, etc.), explore specific category vouchers
- Offer stronger incentives selectively to historically valuable customers.
- Use lower-cost CRM channels such as email reminders for lower-value at-risk customers.
- Monitor whether reactivated customers make another purchase after receiving the campaign.
- Can potentially explore A/B testing: Targeted incentives/messaging vs. standard win-back message.

### Recent / Low Frequency
#### Strategy: Encourage Repeat Purchase

Recent / Low Frequency customers represent 11.87% of customers and contribute 4.74% of total revenue. These customers have purchased recently but have relatively low purchase frequency, presenting an opportunity to encourage a second or subsequent purchase.

**Recommended actions:**
- Post-purchase comms encouraging repeat purchases.
- Provide personalised product recommendations based on their recent purchase.
- Use time-limited return vouchers or second-purchase incentives.
- Introduce customers to loyalty benefits to encourage longer-term engagement.
- Introduce cashback/return vouchers (note not to utilise too much marketing budget on this as this customer segment is not as valuable).

### Inactive / Low Frequency
#### Strategy: Low-Cost Reactivation (only if budget allows)

Inactive / Low Frequency customers make up 30.77% of the customer base but contribute only 7.67% of total revenue. Given their relatively low historical value and lack of recent activity, costly reactivation incentives may not be justified for the entire segment.

**Recommended actions:**
- Use low-cost CRM channels such as email for reactivation campaigns.
- Promote relevant new products or seasonal campaigns to encourage customers to return.
- Test small incentives before offering larger discounts.
- Prioritise marketing spend towards customers who demonstrate renewed engagement.

### Regular
#### Strategy: Grow Customer Value

Regular customers form the largest segment at 36.72% of customers and contribute 23.55% of total revenue. These customers demonstrate moderate purchasing behaviour and represent an opportunity to grow future customer value.

**Recommended actions:**
- Use personalised recommendations and cross-selling to increase basket size.
- Encourage more frequent purchases through targeted CRM campaigns.
- Promote loyalty programme benefits or spending milestones.
- Identify customers whose behaviour is approaching High-Value Loyal status and target them with relevant incentives.

In [85]:
rfm.to_csv("rfm_customer_segments.csv")